phpstan.dist.neon

```
@@ -13,4 +13,24 @@ parameters:
         -
             identifier: constructor.unusedParameter
             paths:
+                - src/Behat/Behat/Definition/Printer/ConsoleDefinitionPrinter.php
```

All these classes have an unused parameter in the constructor. We cannot remove it because they are handled in a generic way that provides this parameter when constructing them so we need to ignore the error

---

Would it be better / more obvious to introduce a new interface(s) with the expected constructor signature for these classes?

---

Actually, looking into this in more detail I can see that almost all these classes are never constructed directly other than as part of the container. Should we just remove these unused constructor parameters? There is the tiny possibility that someone out there is constructing them directly but my guess is that all code will be retrieving them from the container so removing the unused parameter should have no effect. What do you think?

The exception is the `ReturnTypeTransformation` class, which is not constructed as part of the container and which is constructed in the `TransformationCalleeFactory` `create()` method. The constructor for these transformations are called in a generic way assuming that they implement the `SimpleArgumentTransformation` interface. We could maybe add the constructor signature to this interface but this would not help in removing the fact that the constructor for this particular class does not use one of the parameters, so I think we need to keep the error ignore

---

once an interface enforces the signature, phpstan should not be complaining about the unused parameter as it will know that the interface requires it.

---

@stof I tried this but after adding the constructor signature to the interface, PHPStan is still complaining about the unused constructor parameter and in fact this is listed as a bug in PHPStan 😞 
https://github.com/phpstan/phpstan/issues/11454

---

I created a PR in PHPStan to fix this issue https://github.com/phpstan/phpstan-src/pull/3777 😄

---

> Actually, looking into this in more detail I can see that almost all these classes are never constructed directly other than as part of the container. Should we just remove these unused constructor parameters? There is the tiny possibility that someone out there is constructing them directly but my guess is that all code will be retrieving them from the container so removing the unused parameter should have no effect. What do you think?

Hmmm - I think you are probably right that it *should* be safe to remove them, I guess even if someone has an extension that is wrapping/decorating them they should really be configuring that through the container as well. But it does make me a little nervous...

---

OK, so I left the PHPStan error exclusions and added a deprecated warning to these parameters. I also added this to the list of things to do for our next major release. Does everything look OK now?